nvflare/app_common/xgb/sec/processor_data_converter.py

```diff
@@ -0,0 +1,137 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Dict, List, Tuple
+
+from nvflare.apis.fl_context import FLContext
+from nvflare.app_common.xgb.sec.dam import DamDecoder, DamEncoder
+from nvflare.app_common.xgb.sec.data_converter import (
+    AggregationContext,
+    DataConverter,
+    FeatureAggregationResult,
+    FeatureContext,
+)
+
+DATA_SET_GH_PAIRS = 1
+DATA_SET_AGGREGATION = 2
+DATA_SET_AGGREGATION_WITH_FEATURES = 3
+DATA_SET_AGGREGATION_RESULT = 4
+
+SCALE_FACTOR = 1000000.0  # Preserve 6 decimal places
+
+
+class ProcessorDataConverter(DataConverter):
+    def __init__(self):
+        super().__init__()
+        self.features = []
+        self.feature_list = None
+        self.num_samples = 0
+
+    def decode_gh_pairs(self, buffer: bytes, fl_ctx: FLContext) -> List[Tuple[int, int]]:
+        decoder = DamDecoder(buffer)
+        if not decoder.is_valid():
+            raise RuntimeError("GH Buffer is not properly encoded")
+
+        if decoder.get_data_set_id() != DATA_SET_GH_PAIRS:
+            raise RuntimeError(f"Data is not for GH Pairs: {decoder.get_data_set_id()}")
+
+        float_array = decoder.decode_float_array()
+        result = []
+        self.num_samples = int(len(float_array) / 2)
+
+        for i in range(self.num_samples):
+            result.append((self.float_to_int(float_array[2 * i]), self.float_to_int(float_array[2 * i + 1])))
+
+        return result
+
+    def decode_aggregation_context(self, buffer: bytes, fl_ctx: FLContext) -> AggregationContext:
+        decoder = DamDecoder(buffer)
+        if not decoder.is_valid():
+            raise RuntimeError("Aggregation Buffer is not properly encoded")
+        data_set_id = decoder.get_data_set_id()
+        cuts = decoder.decode_int_array()
+
+        if data_set_id == DATA_SET_AGGREGATION_WITH_FEATURES:
+            self.feature_list = decoder.decode_int_array()
+            num = len(self.feature_list)
+            slots = decoder.decode_int_array()
+            for i in range(num):
+                bin_assignment = []
+                for row_id in range(self.num_samples):
+                    _, bin_num = self.slot_to_bin(cuts, slots[row_id * num + i])
+                    bin_assignment.append(bin_num)
+
+                bin_size = self.get_bin_size(cuts, self.feature_list[i])
+                feature_ctx = FeatureContext(self.feature_list[i], bin_assignment, bin_size)
+                self.features.append(feature_ctx)
+        elif data_set_id != DATA_SET_AGGREGATION:
+            raise RuntimeError(f"Invalid DataSet: {data_set_id}")
+
+        node_list = decoder.decode_int_array()
+        sample_groups = {}
+        for node in node_list:
+            row_ids = decoder.decode_int_array()
+            sample_groups[node] = row_ids
+
+        return AggregationContext(self.features, sample_groups)
```

Comment:
The whole method should be in try/except block, since any decode method call could fail.